android-sdk/README.md

```diff
@@ -0,0 +1,93 @@
+
+The Core SDK will provide a request lifecycle that the Psuedocode section will demonstrate.
+
+### Service Modules
+
+Service modules will each represent a singular service that is supportd by mobile-core.  We will use, where available and practical, existing libraries for services and in those cases the modules will provide the necessary implementation of Core APIs to make them compatable with the Core SDK lifecycles and features.  The first Services we are looking at integrations will be Keycloak, fh-sync-service, and a RESTful API service module.
+
+### Use cases and psuedo code examples
+
+#### Defining a Service Module
+A service interface will be responsible for building requests, parsing responses, handling errors.  Services may depend on each other and the method of service composition is currently under evaluation. However, services will have a resolution order where a request is passed along a chain until it is ready to be executed.
+#### Create a Request
+A request is a object that is ready to be executed. The request contains in itself any authentication tokens, headers, etc necessary to connect to a service running in a mobile-core environment.
+For this example, let’s say the user is writing a photo sharing application.  They will probably in their application have an interface that looks like this : 
+```
+public interface UploadService {
+   public RequestHandle upload(Picture mypicture);
+}
+```
+There will be, in OpenShift, a Mobile application, a sync server, a keycloak server, and a storage server.  An implementation of this interface could load a keycloak token from the device, validate the token, start a new OkHttpRequest object, Upload the file etc etc.  What I am getting at is that this is a very laborious process.  However, let’s annotate this call a bit.
+```
+@MobileService(requires = {KeycloakService.class, SyncService.class, StorageService.class},
+                  appKey=”xyz”)
+public interface UploadService {
+   @ServiceMethod
+   public RequestHandle upload(Picture mypicture);
+}
+```
+In this world, the `@MobileService` annotation would signal to the core SDK that this interface should be implemented by our libraries and it requires a request to include the Keycloak, Sync, and Storage.
+The code could then be invoked by the developer as follows : 
+```
+RequestHandle handle = org.aerogear.android.core.MobileService.from(UploadService.class).upload(myPicture);
```

Comment:
I am not entirely sure how this works. More specifically, I am not sure how `MobileService` will know what methods to invoke in the annotated classes. 

For example, the `KeycloakService` may have multiple methods, e.g. login, logout, isTokenExpired, addTokenHeader etc. In this case, the `MobileService` should invoke the `addTokenHeader` method. But how does it will know which one to invoke?

Reply:
I think it may need more code.  We may do couple different approaches for SDK (as prototypes) and pick best patterns.

Reply:
Dependency injection, either with compile time or run time classpath scanning will bridge annotated services and the MobileCore endpoints.

Something we need to work out is the idea of a request and a response lifecycle.  In AeroGear Android Pipe this was done using PipeModules which defined a few methods that are called during a request/response.  Pipe also had mechanisms for controlling how requests and responses were written to and parsed from HTTP services.

Links : 
--Modules
Pipe Module : https://github.com/aerogear/aerogear-android-pipe/blob/master/library/src/main/java/org/jboss/aerogear/android/pipe/module/PipeModule.java
Pipe Module Implementation (OAuth2) : https://github.com/aerogear/aerogear-android-authz/blob/master/library/src/main/java/org/jboss/aerogear/android/authorization/oauth2/OAuth2AuthzModule.java

-- Request/Response
RequestBuilder : https://github.com/aerogear/aerogear-android-pipe/blob/master/library/src/main/java/org/jboss/aerogear/android/pipe/RequestBuilder.java
Response Parser : https://github.com/aerogear/aerogear-android-pipe/blob/master/library/src/main/java/org/jboss/aerogear/android/pipe/ResponseParser.java

Example app showing all of this working together : https://github.com/aerogear/aerogear-android-cookbook/blob/master/ShootAndShare/app/src/main/java/org/jboss/aerogear/android/cookbook/shootandshare/util/GooglePlusHelper.java

Reply:
This code means I can do what the anotation process will do by my self?

Reply:
Yes.  The goal is to make annotations be the golden scenario and work for most use cases.

Reply:
@secondsun it's still not clear to me how the `MobileService` implementation can find the right method to invoke. Are you saying that the methods in `KeycloakService` and `SnycService` will be annotated? Do you have a little bit more examples to explain how this works?

Reply:
@wei-lee Right now my thought is either the service and its methods will be an annotated POJO, or the service will be annotated at the class level and implement a interface for the methods.